### PR TITLE
Add watching on Secrets in drcluster_controller

### PR DESCRIPTION
**Pull request by ELENAGER**
Upon Hub recovery, it takes time till secrets are copied. While secrets are still not copied, drcluster_controller fails to find them several times and goes to exponential backoff. When secrets are finally copied, drcluster_controller is not notified, because it was not registered for it. 
Watching on Secrets added

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2248666

### CODE REVIEW

File: controllers/drcluster_controller.go [hunk 1]
```diff
@@ -120,6 +120,10 @@ func (r *DRClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Watches(&source.Kind{Type: &ramen.DRPlacementControl{}}, drpcMapFun, builder.WithPredicates(drpcPred())).
 		Watches(&source.Kind{Type: &ocmworkv1.ManifestWork{}}, mwMapFun, builder.WithPredicates(mwPred)).
 		Watches(&source.Kind{Type: &viewv1beta1.ManagedClusterView{}}, mcvMapFun, builder.WithPredicates(mcvPred)).
+		Watches(&source.Kind{Type: &corev1.Secret{}},
+			handler.EnqueueRequestsFromMapFunc(r.drClusterSecretMapFunc),
+			builder.WithPredicates(util.CreateOrDeleteOrResourceVersionUpdatePredicate{}),
+		).
```
Comment (nirs) [inline controllers/drcluster_controller.go:126]: This is little bit inconsistent with other code - in other watches we have a function (e.g. mcvMapFun) and predicate (e.g. mcvPred). Here we have harder to follow structure.

File: controllers/drcluster_controller.go [hunk 2]
```diff
@@ -141,6 +145,24 @@ func (r *DRClusterReconciler) drClusterConfigMapMapFunc(configMap client.Object)
 	return requests
 }
 
+func (r *DRClusterReconciler) drClusterSecretMapFunc(secret client.Object) []reconcile.Request {
+	if secret.GetNamespace() != NamespaceName() {
```
Comment (nirs) [inline controllers/drcluster_controller.go:149]: Not clear what is this namespace. Not an issue in your change but maybe it is time to make this more clear.

File: controllers/drcluster_controller.go [hunk 3]
```diff
@@ -141,6 +145,24 @@ func (r *DRClusterReconciler) drClusterConfigMapMapFunc(configMap client.Object)
...
+		return []reconcile.Request{}
```
Comment (nirs) [inline controllers/drcluster_controller.go:150]: Why not return nil? (empty slices is nil)

File: controllers/drcluster_controller.go [hunk 4]
```diff
@@ -141,6 +145,24 @@ func (r *DRClusterReconciler) drClusterConfigMapMapFunc(configMap client.Object)
...
+	}
+
+	drcusters := &ramen.DRClusterList{}
```
Comment (nirs) [inline controllers/drcluster_controller.go:153]: We don't need to create the DRClusterList on the heap - we can do:

    drclusters := ramen.DRClusterList{}
    r.Client.List(context.TODO(), &drcusters)

File: controllers/drcluster_controller.go [hunk 5]
```diff
@@ -141,6 +145,24 @@ func (r *DRClusterReconciler) drClusterConfigMapMapFunc(configMap client.Object)
...
+	if err := r.Client.List(context.TODO(), drcusters); err != nil {
+		return []reconcile.Request{}
+	}
+
+	requests := make([]reconcile.Request, len(drcusters.Items))
+	for i, drcluster := range drcusters.Items {
+		requests[i].Name = drcluster.GetName()
```
Comment (nirs) [inline controllers/drcluster_controller.go:160]: Doesn't this loop copy all drclusters from drcusters.Items? We can avoid the copy if we do:

    for i := range drcusters.Items {
        requests[i].Name = drclusters.Items[i].GetName()
    }

**Review by ShyamsundarR — APPROVED** [review]
I was thinking of a envtest and looked for parallels in DRPolicy, which also does not have a test for a secret event (though it is covered in other ways as per the coverage report).

Otherwise this looks good. In case we still want to address the optimizations that nir suggested we can amend it accordingly.

File: controllers/drcluster_controller.go (hunk 1 above)
Comment (ELENAGER) [reply controllers/drcluster_controller.go:126]: Not exactly, [here](https://github.com/RamenDR/ramen/blob/85fa41c3f9b57a62d94b966a9189ea740334ddb2/controllers/drcluster_controller.go#L119) is the same structure

File: controllers/drcluster_controller.go (hunk 3 above)
Comment (ELENAGER) [reply controllers/drcluster_controller.go:150]: I am returning it for the consistency with the other code. The same return can be found in 5 more places in this file.

File: controllers/drcluster_controller.go (hunk 4 above)
Comment (ELENAGER) [reply controllers/drcluster_controller.go:153]: You are probably right, but it's also for consistency. We are always creating lists on heap (I've looked through all our code).

File: controllers/drcluster_controller.go (hunk 5 above)
Comment (ELENAGER) [reply controllers/drcluster_controller.go]: Done

File: controllers/drcluster_controller.go [hunk 6]
```diff
@@ -141,6 +145,24 @@ func (r *DRClusterReconciler) drClusterConfigMapMapFunc(configMap client.Object)
...
+	if secret.GetNamespace() != NamespaceName() {
+		return []reconcile.Request{}
+	}
+
+	drcusters := &ramen.DRClusterList{}
+	if err := r.Client.List(context.TODO(), drcusters); err != nil {
+		return []reconcile.Request{}
+	}
+
+	requests := make([]reconcile.Request, len(drcusters.Items))
```
[the remark this answers is not in the document]
Comment (ELENAGER) [reply controllers/drcluster_controller.go:157]: I added [an optimization issue](https://github.com/RamenDR/ramen/issues/1170)

**Review by ShyamsundarR — APPROVED** [review]

**Pull request merged**